pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,463 @@
+			continue
+		}
+
+		// Skip resource kinds that are not referenced in CSV.
+		if !s.scheme.Recognizes(bufObj.GroupVersionKind()) {
+			continue
+		}
+
+		if bufObj.GetKind() == CSVKind {
+			if s.existingCSVPath != "" {
+				return fmt.Errorf("more than one ClusterServiceVersion is found in bundle")
+			}
+			s.existingCSVPath = path
+		}
+		s.appendTostructuredObjectMap(bufObj.GetKind(), bufObj)
+	}
+
+	metafiles, err := ioutil.ReadDir(s.metadataDir)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range metafiles {
+		path := filepath.Join(s.metadataDir, file.Name())
+		if err := s.loadNonInferableCSV(path); err != nil {
+			return err
+		}
+		if err := s.loadDependencies(path); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// loadNonInferableCSV loads the nonInferableCSV in the synthesize structure and puts all descriptions.
+// Function return nil if file is not a parsable to nonInferableCSV.
+// Function returns error on duplicated CRD GVK based on the descriptions or parsing error.
+func (s *synthesize) loadNonInferableCSV(path string) error {
+	reader, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	csvBasics := &NonInferableCSV{}
+	decoder := yaml.NewYAMLOrJSONDecoder(reader, 30)
+	if err = decoder.Decode(csvBasics); err != nil || csvBasics.Name == "" {
+		return nil
+	}
+
+	if s.nonInferableCSV.Name != "" {
+		return fmt.Errorf("non-inferable CSV information already exist")
+	}
+	s.nonInferableCSV = csvBasics
+
+	for _, apiDes := range s.nonInferableCSV.ApiDescriptions {
+		for _, descriptor := range apiDes.Descriptors {
+			if descriptor.DescriptorType != StatusDescriptorType && descriptor.
+				DescriptorType != SpecDescriptorType && descriptor.DescriptorType != ActionDescriptorType {
+				return fmt.Errorf("unrecognized `DescriptorType`, please use `status`, `spec`, or `action`")
+			}
+		}
+
+		gvk, err := apiDes.GetGroupVersionKind()
+		if err != nil {
+			return err
+		}
+		if _, ok := s.apiDescriptionMap[gvk]; ok {
+			return fmt.Errorf("descriptors with duplicated GVK %v are found", gvk)
+		}
+		api := apiDes
+		s.apiDescriptionMap[gvk] = &api
+	}
+
+	return nil
+}
+
+// appendTostructuredObjectMap converts the unstructured object and try to convert it into a specified interface
+// before appending to the unstructuredObjectMap.
+func (s *synthesize) appendTostructuredObjectMap(kind string, bufObj unstructured.Unstructured) {
+	s.unstructuredObjectMap[kind] = append(s.unstructuredObjectMap[kind], bufObj)
+}
+
+// loadDependencies tries to decode an object into registry.DependenciesFile{}.
+// It updates the dependencies in synthesize once decode is successful.
+func (s *synthesize) loadDependencies(path string) error {
+	dependenciesFile := registry.DependenciesFile{}
+	if err := registry.ParseDependenciesFile(path, &dependenciesFile); err != nil {
+		if strings.HasPrefix(err.Error(), "Unable to decode the dependencies file") {
```

Comment:
It would be a good idea to have a more programmatic way to interpret the error values returned by ParseDependenciesFile. String comparison is more brittle than something like a sentinel value or a predicate function.